libs/nss.c

```diff
@@ -27,6 +27,31 @@ int _nss_tcb_endspent(void)
 	return 1;
 }
 
+/* IEEE Std 1003.1-2001 allows only the following characters to appear
+   in group- and usernames: letters, digits, underscores, periods,
+   <at>-signs (@), and dashes.   The name may not start with a dash or @.
+   The "$" sign is allowed at the end of usernames to allow typical
+   Samba machine accounts.  Regex: ^[a-z_.][a-z0-9_@.-]*[$]$ */
+static int
+is_valid_username (const char *un)
+{
+	if (!un || !strlen(un) || un[0] == '-' || un[0] == '@' ||
+	    /* dirsep || curdir || parentdir */
+	    strchr(un, '/') || !strcmp(un, ".") || !strcmp(un, ".."))
+		return 0;
+
+	int retval = 1;
+	size_t unlen = strlen(un);
+
+	for (size_t i = 0; !retval && i < unlen; i++) {
+	char c = un[i];
```

Comment:
Wrong indentation.

Also, instead of the index and length, I'd use a pointer. The last character check would easily become just a check of `!*p && c == '$'` after `p` has been advanced by a prior `char c = *p++;`.

Reply:
Adapted.